keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/VersioningStrategy.kt

```diff
@@ -0,0 +1,148 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package com.netflix.spinnaker.keel.api
+
+import com.fasterxml.jackson.annotation.JsonIgnore
+import com.fasterxml.jackson.databind.JsonDeserializer
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize
+import com.netflix.frigga.ami.AppVersion
+import com.netflix.rocket.semver.shaded.DebianVersionComparator
+import com.netflix.spinnaker.keel.api.SortType.INCREASING
+import com.netflix.spinnaker.keel.api.SortType.SEMVER
+import com.netflix.spinnaker.keel.serialization.PropertyNamePolymorphicDeserializer
+import net.swiftzer.semver.SemVer
+import org.slf4j.LoggerFactory
+import org.springframework.util.comparator.NullSafeComparator
+
+/**
+ * Strategy for how to sort versions of artifacts.
+ */
+@JsonDeserialize(using = VersioningStrategyDeserializer::class)
+sealed class VersioningStrategy(
+  @JsonIgnore
+  open val comparator: Comparator<String>
+)
+
+@JsonDeserialize(using = JsonDeserializer.None::class)
+object DebianSemVerVersioningStrategy : VersioningStrategy(DEBIAN_VERSION_COMPARATOR) {
+  override fun toString(): String =
+    javaClass.simpleName
+
+  override fun equals(other: Any?): Boolean {
+    return other is DebianSemVerVersioningStrategy
+  }
+}
+
+@JsonDeserialize(using = JsonDeserializer.None::class)
+data class DockerVersioningStrategy(
+  val strategy: TagVersionStrategy,
+  val captureGroupRegex: String? = null
+) : VersioningStrategy(TagComparator(strategy, captureGroupRegex)) {
+  override fun toString(): String =
+    "${javaClass.simpleName}[strategy=$strategy, captureGroupRegex=$captureGroupRegex]}"
+}
+
+/**
+ * Comparator that supports all the tag options for docker containers
+ */
+class TagComparator(
+  private val strategy: TagVersionStrategy,
+  private val customRegex: String? = null
+) : Comparator<String> {
+
+  private val semverComparator = NullSafeComparator(SEMVER_COMPARATOR, false)
+  private val increasingComparator = NullSafeComparator(INCREASING_COMPARATOR, false)
+
+  override fun compare(o1: String, o2: String): Int {
+    val i1 = parseWithRegex(o1, strategy, customRegex)
+    val i2 = parseWithRegex(o2, strategy, customRegex)
+    return when (strategy.sortType) {
+      SEMVER -> semverComparator.compare(parseSemver(i1), parseSemver(i2))
+      INCREASING -> increasingComparator.compare(i1?.toIntOrNull(), i2?.toIntOrNull())
+    }
+  }
+
+  companion object {
+    private val log by lazy { LoggerFactory.getLogger(this::class.java) }
+
+    fun parseWithRegex(input: String, strategy: TagVersionStrategy, customRegex: String?): String? {
+      val regex = Regex(customRegex ?: strategy.regex)
+      val result = regex.find(input) ?: return null
+      return when (result.groupValues.size) {
+        2 -> result.groupValues[1]
+        1 -> {
+          log.warn("Regex (${customRegex ?: strategy.regex}) produced zero capture groups on tag $input")
```

Comment:
Could it be that there was just no capturing groups in the custom regex and so group 0 (entire match) should be returned here?

Reply:
I don't think so. For example, what if the capture group for docker is meant to capture the job number out of a tag like this `master-h5.5a52206` (regex: `^master-h(\d+).*$` will capture `5` from `h5`). Also present in the list of tags is `latest`. That regex will not produce a capture group on `latest`, so we want to exclude it from the sort.